qlib/kernel/Kernel_cc.rs

```diff
@@ -0,0 +1,2085 @@
+// Copyright (c) 2021 Quark Container Authors / 2018 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::mem::size_of;
+
+use self::range::Range;
+use alloc::boxed::Box;
+
+use super::super::common::*;
+use super::super::config::*;
+use super::super::linux_def::*;
+use super::super::loader::*;
+use super::super::qmsg;
+use super::super::qmsg::*;
+use super::super::socket_buf::*;
+use super::super::tsot_msg::*;
+use super::super::*;
+use super::Kernel::*;
+use crate::kernel_def::HyperCall64;
+use crate::qlib::control_msg::ControlMsg;
+use crate::qlib::nvproxy::frontend_type::RMAPIVersion;
+use crate::qlib::proxy::*;
+use crate::GLOBAL_ALLOCATOR;
+use crate::GUEST_HOST_SHARED_ALLOCATOR;
+
+extern "C" {
+    pub fn rdtsc() -> i64;
+}
+
+impl HostSpace {
+    //Here exists memory leak, vec in shared struct cannot be dealloc
+    pub fn LoadProcessKernel_cc(processAddr: u64) -> i64 {
+        let mut new_process = Box::new_in(Process::default(), GUEST_HOST_SHARED_ALLOCATOR);
+        let process_ptr = &mut *new_process as *mut _;
+        let mut msg = Box::new_in(
+            Msg::LoadProcessKernel(LoadProcessKernel {
+                processAddr: process_ptr as u64,
+            }),
+            GUEST_HOST_SHARED_ALLOCATOR,
+        );
+        let ret = HostSpace::Call(&mut msg, false) as i64;
+        let private_process = unsafe { &mut *(processAddr as *mut Process) };
+        private_process.clone_from_shared(process_ptr);
+        //info!("###Cloned Process:{:#?}\n",private_process);
+        //Here could have memory leak since
```

Comment:
Is there still a memory leak? Shouldn't the shared memory be dallocated in Global_Allocator when the fuction drops new_process?

Reply:
Yes, the memory leak should not exist here now.